### PR TITLE
Evangelos Asimakopoulos code GWI

### PR DESCRIPTION
**Pull request by evangelosas**
The code was written mostly by Junie Pro and my role was to supervise and make all the necessary changes in order to provide a deliverable of high standard. The time gained gave me the opportunity to add more tests and functionality.

### CODE REVIEW

File: Containerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# syntax=docker/dockerfile:1
+
+# Podman uses the same Dockerfile syntax. This Containerfile mirrors the Dockerfile.
+
+# ----- Build stage -----
+FROM golang:1.22-alpine AS builder
```
Comment (alexandros-georgantas) [inline Containerfile:6]: Could you please explain your rationale behind keeping this old version of golang?

File: README.md [hunk 2]
```diff
@@ -1,31 +1,140 @@
...
+  - Body: `{"description":"new description"}`
+- DELETE `/users/{userID}/favourites/{assetID}`
+
+Asset response objects contain common fields `id`, `type`, `description`, `createdAt` plus type-specific fields.
+
+## Storage
+
+Two store implementations:
+- In-memory (default): fast, volatile.
+- File-backed (JSON): persists to a single file. Configure via env vars:
```
Comment (alexandros-georgantas) [inline README.md:60]: I understand your choice of using an in-memory store for the scope of this assignment, however I would like to hear your thoughts about the file backed option.
For your file-based storage why you haven't chosen `sqlite` for example which would have given you some structure to work with?

File: gwiexercise/decodeassets.go
Comment (alexandros-georgantas) [inline gwiexercise/decodeassets.go:1]: Could you please review this file and suggest if it could be simplified?

File: gwiexercise/decodeassets.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+type AddAssetRequest struct {
+	ID          string          `json:"id"`
+	Type        AssetType       `json:"type"`
+	Description string          `json:"description"`
+	Payload     json.RawMessage `json:"payload"`
+}
+
+func (chart *ChartAsset) MarshalJSON() ([]byte, error) {
+	type alias ChartAsset
```
Comment (alexandros-georgantas) [inline gwiexercise/decodeassets.go:17]: Could you please explain the need of type alias and what you are trying to solve?

File: gwiexercise/server.go
Comment (alexandros-georgantas) [inline gwiexercise/server.go:1]: Could you please suggest potential refactoring of this file as it scores low when it comes to structure and readability?

File: gwiexercise/server.go [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+			return
+		}
+		assetID := segs[3]
+		s.handleUserFavouriteByID(w, r, userID, assetID)
+		return
+	}
+	writeError(w, http.StatusNotFound, "not found")
+}
+
+func (s *Server) handleUserFavourites(w http.ResponseWriter, r *http.Request, userID string) {
```
Comment (alexandros-georgantas) [inline gwiexercise/server.go:92]: What is you opinion on having a single handler handling all the HTTP methods?

File: main.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+		if err != nil {
+			log.Fatalf("failed to init file store: %v", err)
+		}
+		store = fs
+		log.Printf("Using file store at %s", path)
+	} else {
+		store = gwi.NewInMemoryStore()
+		log.Printf("Using in-memory store")
+	}
+	server := gwi.NewServer(store)
```
Comment (alexandros-georgantas) [inline main.go:28]: Could you suggest a more graceful approach of initializing the HTTP server?

**Review by alexandros-georgantas — COMMENTED** [review]
Hello Evangelos 👋 ,

Thank you for taking the time working on this assignment!
I've left some comments as part of my review of this PR and I would love to hear your thoughts!

Thanks,
Alex

File: gwiexercise/decodeassets.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+type AddAssetRequest struct {
+	ID          string          `json:"id"`
+	Type        AssetType       `json:"type"`
+	Description string          `json:"description"`
+	Payload     json.RawMessage `json:"payload"`
```
Comment (alesr) [inline gwiexercise/decodeassets.go:13]: Is there a way to define acceptable payload types at compile time instead of just accepting raw data?

File: gwiexercise/decodeassets_test.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+	}
+}
+
+func TestDecodeAssetFromAddRequest(t *testing.T) {
+	tests := []struct {
+		name       string
+		req        AddAssetRequest
+		wantType   AssetType
+		wantErr    bool
+		errorMsg   string
```
Comment (alesr) [inline gwiexercise/decodeassets_test.go:60]: In your test, are you interested in verifying whether the implementation correctly returns an error when it should, or whether the specific error message returned is the expected one?

Why do you think, in this particular case, the error string itself matters?

File: temporary.txt [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+   - Ensure Content-Type is application/json on POST; return 415 if not.
+   - The handler already sets Allow for non-POST; keep tests verifying 405 and Allow header.
+
+8) Logging and observability
+   - Log batch size, user ID (redacted as necessary), and error categories to aid debugging and monitoring.
+
+9) Test coverage
+   - Add unit tests that assert: (a) status mapping without relying on string matching (once typed errors exist), (b) duplicate IDs within a batch, (c) duplicate against existing store, (d) large offset/limit interaction after bulk insert, (e) body Content-Type handling, (f) maximum batch size enforcement.
+
+These improvements would make the bulk endpoint more robust, maintainable, and user-friendly without changing the core behavior clients rely on today.
```
Comment (alesr) [inline temporary.txt:68]: Could you please explain what is this file for?
